m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+
+    private static final int THREAD_POOL_SIZE = 10;
+
+    private static final String SERVICE_TAG = "service";
+    private static final String ENV_TAG = "env";
+    private static final String HOST_TAG = "host";
+
+    private static final String DEFAULT_HISTOGRAM_BUCKET_ID_NAME = "bucketid";
+    private static final String DEFAULT_HISTOGRAM_BUCKET_NAME = "bucket";
+    private static final int DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION = 6;
+
+    private static final int EMIT_METRIC_BATCH_OVERHEAD = 19;
+    private static final int MIN_METRIC_BUCKET_ID_TAG_LENGTH = 4;
+
+    private M3.Client client;
+
+    private final MetricBatch metricBatch;
+
+    private final Object calcLock = new Object();
+    private TCalcTransport calc;
+    private TProtocol calcProtocol;
+
+    private int freeBytes;
+
+    private ResourcePool resourcePool;
+
+    private String bucketIdTagName;
+    private String bucketTagName;
+    private String bucketValFmt;
+
+    // Holds metrics to be flushed
+    private BlockingQueue<SizedMetric> metricQueue;
+    // The service used to execute metric flushes
+    private ExecutorService executor;
+    // Used to keep track of how many processors are in progress in
+    // order to wait for them to finish before closing
+    private Phaser phaser = new Phaser(1);
+
+    // Use inner Builder class to construct an M3Reporter
+    private M3Reporter(Builder builder) {
+        try {
+            // Builder verifies non-null, non-empty hostPorts
+            String[] hostPorts = builder.hostPorts;
+
+            TProtocolFactory protocolFactory;
+            if (builder.protocol == ThriftProtocol.COMPACT) {
+                protocolFactory = new TCompactProtocol.Factory();
+            } else {
+                protocolFactory = new TBinaryProtocol.Factory();
+            }
+
+            TTransport transport = new TUdpClient(hostPorts[0]);
+
+            client = new M3.Client(protocolFactory.getProtocol(transport));
+            resourcePool = new ResourcePool(builder.maxQueueSize, protocolFactory);
+
+            Set<MetricTag> commonTags = addMetricTagsToSet(builder.commonTags);
+
+            // Set and ensure required tags
+            if (!builder.commonTags.containsKey(SERVICE_TAG)) {
+                if (builder.service == null || builder.service.isEmpty()) {
+                    throw new IllegalArgumentException(String.format("%s common tag is required", SERVICE_TAG));
+                }
+
+                commonTags.add(createMetricTag(SERVICE_TAG, builder.service));
+            }
+            if (!builder.commonTags.containsKey(ENV_TAG)) {
+                if (builder.env == null || builder.env.isEmpty()) {
+                    throw new IllegalArgumentException(String.format("%s common tag is required", ENV_TAG));
+                }
+
+                commonTags.add(createMetricTag(ENV_TAG, builder.env));
+            }
+            if (builder.includeHost) {
```

Comment:
nit: simplify to `if (builder.includeHost && !builder.commonTags.containsKey(HOST_TAG)) { ... `